mixer/test/listbackend/list.go

```diff
@@ -0,0 +1,190 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// nolint:lll
+//go:generate go run $GOPATH/src/istio.io/istio/mixer/tools/mixgen/main.go adapter -n listbackend-nosession -c $GOPATH/src/istio.io/istio/mixer/adapter/list/config/config.proto_descriptor -s=false -t listentry -o nosession.yaml
+
+package listbackend
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net"
+	"sync"
+
+	"google.golang.org/grpc"
+
+	adptModel "istio.io/api/mixer/adapter/model/v1beta1"
+	"istio.io/istio/mixer/adapter/list"
+	"istio.io/istio/mixer/adapter/list/config"
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/pkg/pool"
+	"istio.io/istio/mixer/pkg/runtime/handler"
+	"istio.io/istio/mixer/template/listentry"
+)
+
+type (
+	// Server is basic server interface
+	Server interface {
+		Addr() net.Addr
+		Close() error
+		Run()
+		Wait() error
+	}
+
+	// nosessionServer models no session adapter backend.
+	nosessionServer struct {
+		listener net.Listener
+		shutdown chan error
+		server   *grpc.Server
+
+		rawcfg      []byte
+		builder     adapter.HandlerBuilder
+		env         adapter.Env
+		builderLock sync.RWMutex
+
+		h listentry.Handler
+	}
+)
+
+var _ listentry.HandleListEntryServiceServer = &nosessionServer{}
+
+func (s *nosessionServer) getHandler(rawcfg []byte) (listentry.Handler, error) {
+	s.builderLock.RLock()
+	if 0 == bytes.Compare(rawcfg, s.rawcfg) {
+		h := s.h
+		s.builderLock.RUnlock()
+		return h, nil
+	}
+	s.builderLock.RUnlock()
+
+	// establish session
+	cfg := &config.Params{}
+	if err := cfg.Unmarshal(rawcfg); err != nil {
+		return nil, err
+	}
+
+	s.builderLock.Lock()
+	defer s.builderLock.Unlock()
+
+	if 0 == bytes.Compare(rawcfg, s.rawcfg) {
+		return s.h, nil
+	}
+
+	s.builder.SetAdapterConfig(cfg)
+	if ce := s.builder.Validate(); ce != nil {
+		return nil, ce
+	}
+
+	h, err := s.builder.Build(context.Background(), s.env)
+	if err != nil {
+		return nil, err
+	}
+	s.rawcfg = rawcfg
+	s.h = h.(listentry.Handler)
+	return s.h, err
+}
+
+func instance(in *listentry.InstanceMsg) *listentry.Instance {
+	return &listentry.Instance{
+		Name:  in.Name,
+		Value: in.Value,
+	}
+}
+
+// HandleListEntry records listrequest and responds with the programmed response
+func (s *nosessionServer) HandleListEntry(ctx context.Context, r *listentry.HandleListEntryRequest) (*adptModel.CheckResult, error) {
+	s.env.Logger().Infof("received '%v'", *r)
+	var h listentry.Handler
+	var err error
+	if r.AdapterConfig != nil {
+		h, err = s.getHandler(r.AdapterConfig.Value)
+	} else {
+		h, err = s.getHandler(nil)
```

Comment:
When does this happen? What does it mean to get no config?

Reply:
Yes.. if the input config is empty